rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/StreamT.scala

```diff
@@ -1,39 +1,105 @@
 package coop.rchain.rholang.interpreter.matcher
-import cats.{Applicative, Functor, Monad}
+import cats.implicits._
+import cats.{Monad, MonoidK}
+import coop.rchain.catscontrib.MonadTrans
+import coop.rchain.rholang.interpreter.matcher.StreamT.{SCons, SNil, Step}
 
 import scala.collection.immutable.Stream
-import cats.implicits._
+import scala.util.{Left, Right}
+
+/**
+  * Shamelessly transcribed minimal version of Gabriel Gonzalez's beginner-friendly ListT
+  * https://github.com/Gabriel439/Haskell-List-Transformer-Library/blob/e9a1d19/src/List/Transformer.hs
+  *
+  * See also: http://www.haskellforall.com/2016/07/list-transformer-beginner-friendly-listt.html
+  *
+  * The monads resulting from applying this transformer are as lazy and stacksafe as the underlying F monad you pass.
```

Comment:
Do we have a code proof of this?

Reply:
The laziness is checked by the test you've commented on. Thanks for your comment - I thought there is a stacksafety test in cats suite, but it's just for the tailRecM. I'll add a test for map/flatmap/combine

Reply:
Done